Add Web Components and LIT versions of Complex-DOM TodoMVC

lpardosixtosMs opened the pull request:
This PR adds the Web Components and LIT versions of the Complex-DOM todoMVC workload.

Changes to the complex DOM generator:

- Added constructable stylesheets under the utils folder to be included by the LIT and WC complex todoMVC.

Changes to the standalone versions.

- Added the `show-priority` class name to the `todo-list` elements.
- Added the `data-priority` attribute to the `todo-item` elements.
- The list and list items components now query if the global variables `window.extraTodoListCssToAdopt` and `window.extraCssToAdopt` are populated and adopt the stylesheets stored in them. The variables should not be defined on the standalone version, and they are populated in the complex version from another script.

This PR **DOESN'T** modify which workloads run in the benchmark. That will be done in a follow up PR.

Workload design: https://docs.google.com/document/d/1QoO1VzFf3PZY-lsHdHFKGX6m1p393iEqjHQOzmU0NUI/edit?usp=sharing

Hosted version: https://lpardosixtosMs.github.io/SpeedometerWebComponents/?suite=TodoMVC-WebComponents,TodoMVC-WebComponents-Complex,TodoMVC-LIT,TodoMVC-LIT-Complex-DOM

lpardosixtosMs commented:
@flashdesignory, can you take a look? In order for the item component to adopt the generated stylesheets we need to add some logic in the component's JavaScript. We still don't know which one is the best way to do it, we could:

1. Add logic to import a list of constructable stylesheets from another file. This file will have an empty object in the standalone version and the generated array in the complex version. The content of the file will be replaced in the complex version's build script. (Currently implemented).
2. Replace the `dist/components/todo-item/todo-item.component.js` in the complex version with one containing all the logic we want.
3. Include the generated stylesheets in both standalone and complex versions and use them only when a global flag is present, maybe something in the window object that we can add directly in the html via the build script.

Options 1 and 2 work well for web components but are harder to implement on LIT, since it only produces 2 files in the dist folder. Option 3 works for all but requires us to import the generated stylesheets in both versions and have a global flag. Do you have a preference for any of these options (or any other)?

lpardosixtosMs commented:
@flashdesignory Please let me know if this approach looks good, this should work for the LIT implementation too, which we'll clean up tomorrow.

flashdesignory commented on resources/todomvc/vanilla-examples/javascript-web-components/dist/components/todo-item/todo-item.component.js:
```diff
@@ -146,8 +157,15 @@ class TodoItem extends HTMLElement {
+        if (!EXTRA_CSS_TO_ADOPT)
+            return;
+        const styleSheetIndex = this.index % EXTRA_CSS_TO_ADOPT.length;
+        this.shadow.adoptedStyleSheets.push(EXTRA_CSS_TO_ADOPT[styleSheetIndex]);
```
should we check if a stylesheet exists at this index, before pushing?

flashdesignory commented on .eslintignore:
```diff
@@ -32,8 +32,11 @@ package-lock.json
 
 !/resources/todomvc/big-dom-generator
+resources/todomvc/big-dom-generator/angular
+resources/todomvc/big-dom-generator/javascript-web-components
```
why are we ignoring these?

lpardosixtosMs replied:
These are the generated files, I'm trying to produce them formatted but I think that if the formatter ends up changing them, it might be confusing, WDYT?

flashdesignory replied:
That makes sense

flashdesignory commented on .prettierignore:
```diff
@@ -32,8 +32,11 @@ package-lock.json
 
 !/resources/todomvc/big-dom-generator
+resources/todomvc/big-dom-generator/angular
+resources/todomvc/big-dom-generator/javascript-web-components
```
same as above... are we not formatting these?

flashdesignory commented on resources/todomvc/big-dom-generator/javascript-web-components/generated-variables.css:
```diff
@@ -0,0 +1,50 @@
+	--complex-border-right-color-default: rgba(140, 140, 140, 0.5);
+	--complex-border-top-color-default: rgba(140, 140, 140, 0.5);
+	--complex-column-rule-color-default: rgba(140, 140, 140, 0.5);
+	--complex-outline-color-default: rgba(140, 140, 140, 0.5);
```
A lot of these variables have the same value, just different name, it this on purpose?
Are we just trying to assign as many variables as possible?

lpardosixtosMs replied:
The scalability challenge is to test variable resolution in the presence of many inherited variables, so we didn't put a lot of attention on the values, as long as they didn't disrupt the todoMVC style. That said, we can modify the values to make each variable unique.

What was not intended was to have each list item resolve 8 additional css styles, so I changed it so each `li` and `li > div` elements resolve 2 more styles (like in the other complex versions), and increased the number of generated constructable stylesheets.

flashdesignory replied:
Thanks for the explanation!

flashdesignory commented:
@lpardosixtosMs - can you also add scores standalone / complex?

flashdesignory commented:
The index logic breaks - unless it doesn't matter if the index is not unique. 
<img width="708" alt="Screenshot 2023-07-11 at 8 48 32 AM" src="https://github.com/WebKit/Speedometer/assets/372973/8875bfb8-e5c1-4692-b107-3111f7129463">

flashdesignory commented:
index logic breaks:
<img width="609" alt="Screenshot 2023-07-13 at 8 15 14 PM" src="https://github.com/WebKit/Speedometer/assets/372973/5c6e7e4c-e857-451b-b909-5fbd47fb1c5c">

lpardosixtosMs commented on resources/todomvc/vanilla-examples/javascript-web-components/src/components/todo-app/todo-app.component.js:
```diff
@@ -42,6 +42,7 @@ class TodoApp extends HTMLElement {
 
     addItem(event) {
         const { detail: item } = event;
+        item.index = this.#data.length;
```
Overlooked the cleanup on this part, I'll look into this.

lpardosixtosMs replied:
Done.

lpardosixtosMs commented on resources/todomvc/vanilla-examples/javascript-web-components/src/components/todo-item/todo-item.template.js:
```diff
@@ -2,8 +2,8 @@ const template = document.createElement("template");
 template.innerHTML = `
-    <li class="todo-item">
-        <div class="display-todo">
+    <li class="targeted todo-item">
```
Consistency with the other versions and to have the presence of "rules that will be triggered if it wasn't for the shadow root". Although I think any implementation of the shadow root should ignore global rules quickly regardless of the selector and we already have a lot of other global CSS rules from the complex DOM UI, I'm happy to remove it if you don't see value in it.

flashdesignory replied:
got it - let's leave as is, since the other versions already have it.

lpardosixtosMs commented on resources/tests.mjs:
```diff
@@ -149,6 +149,39 @@ Suites.push({
     ],
 });
 
+Suites.push({
```
Here just for easy testing, should be deleted before merging.

lpardosixtosMs commented:
> @lpardosixtosMs - can you also add scores standalone / complex?

Yes, I will post them here once we get the LIT implementation ready.

flashdesignory commented on resources/todomvc/architecture-examples/lit/src/lib/todo-item.ts:
```diff
@@ -131,18 +132,37 @@ export class TodoItem extends LitElement {
+            return;
+        const styleSheetIndex = this.index % EXTRA_CSS_TO_ADOPT.length;
+        const styleSheetToAdopt = EXTRA_CSS_TO_ADOPT[styleSheetIndex];
+        this.shadowRoot?.adoptedStyleSheets.push(styleSheetToAdopt);
```
should we check if a stylesheet exists at this index, before pushing?

lpardosixtosMs replied:
Oops I forgot the if. Done.

lpardosixtosMs commented:
> @lpardosixtosMs - can you also add scores standalone / complex?

Here are the numbers:

browser                                                                       Firefox             Google Chrome                 Safari
version                                                                        114.0.2             114.0.5735.198      16.5.2 (18615.2.9.11.10)
os                                                                     macos 13.4.1 arm64  macos 13.4.1 arm64  macos 13.4.1 arm64
device                                                                  MacBookPro17,1      MacBookPro17,1      MacBookPro17,1
cpu                                                                     Apple M1 8 cores    Apple M1 8 cores    Apple M1 8 cores
TodoMVC-Lit-Complex-DOM/total                       30.24 ± 1.0%        23.81 ± 0.94%          85.5 ± 1.2%
TodoMVC-Lit/total                                                 23.07 ± 3.9%        18.06 ± 0.67%          21.09 ± 2.2%
TodoMVC-WebComponents-Complex/total         37.32 ± 1.8%        26.912 ± 0.22%         69.92 ± 1.1%
TodoMVC-WebComponents/total                         28.17 ± 0.79%       23.96 ± 1.7%           22.55 ± 2.5%
Score                                                                       41.02 ± 1.2%        52.30 ± 0.70%          29.23 ± 1.4%

Sanity check, standalone numbers before this PR.

browser                                                                   Firefox             Google Chrome           Safari
TodoMVC-Lit/total                                               23.18 ± 3.7%        17.44 ± 2.7%        20.91 ± 1.6%
TodoMVC-WebComponents/total                       29.25 ± 1.5%        24.13 ± 1.9%        23.70 ± 3.4%

lpardosixtosMs commented:
> @lpardosixtosMs - can you also add scores standalone / complex?

@flashdesignory here are the numbers:

OS: macos 13.4.1 arm64. Device: MacBookPro17,1. CPU: Apple M1 8 cores. 

TodoMVC-Lit-Complex-DOM/total       
- Firefox: 30.24 ± 1.0% 
- Chrome:   23.81 ± 0.94%
- Safari:  85.5 ± 1.2%

TodoMVC-Lit/total

- Firefox: 23.07 ± 3.9%
- Chrome: 18.06 ± 0.67%
- Safari: 21.09 ± 2.2%

TodoMVC-WebComponents-Complex/total

- Firefox: 37.32 ± 1.8%
- Chrome: 26.912 ± 0.22% 
- Safari 69.92 ± 1.1%

TodoMVC-WebComponents/total
- Firefox: 28.17 ± 0.79%
- Chrome: 23.96 ± 1.7%
- Safari: 22.55 ± 2.5%

Sanity check: Numbers before this PR:

TodoMVC-Lit/total

- Firefox: 23.18 ± 3.7%
- Chrome: 17.44 ± 2.7%
- Safari: 20.91 ± 1.6%

TodoMVC-WebComponents/total
- Firefox: 29.25 ± 1.5%
- Chrome: 24.13 ± 1.9%
- Safari: 23.70 ± 3.4%

issackjohn commented on resources/tests.mjs:
```diff
@@ -149,6 +149,39 @@ Suites.push({
     ],
 });
 
+Suites.push({
+    name: "TodoMVC-WebComponents-Complex",
+    url: "todomvc/vanilla-examples/javascript-web-components-complex/dist/index.html",
+    tags: ["todomvc", "webcomponents"],
+    async prepare(page) {
+        await page.waitForElement("todo-app");
+    },
+    tests: [
+        new BenchmarkTestStep(`Adding${numberOfItemsToAdd}Items`, (page) => {
+            const input = page.querySelector(".new-todo-input", ["todo-app", "todo-topbar"]);
+            for (let i = 0; i < numberOfItemsToAdd; i++) {
+                input.setValue(getTodoText(defaultLanguage, i));
+                input.dispatchEvent("input");
+                input.enter("keyup");
+            }
+        }),
+        new BenchmarkTestStep("CompletingAllItems", (page) => {
+            const items = page.querySelectorAll("todo-item", ["todo-app", "todo-list"]);
+            for (let i = 0; i < numberOfItemsToAdd; i++) {
+                const item = items[i].querySelectorInShadowRoot(".toggle-todo-input");
+                item.click();
+            }
+        }),
+        new BenchmarkTestStep("DeletingAllItems", (page) => {
+            const items = page.querySelectorAll("todo-item", ["todo-app", "todo-list"]);
+            for (let i = 0; i < numberOfItemsToAdd; i++) {
+                const item = items[i].querySelectorInShadowRoot(".remove-todo-button");
+                item.click();
+            }
+        }),
+    ],
+});
+
```
Don't forget to remove this.

lpardosixtosMs commented:
@rniwa @julienw I'm not sure if this qualifies as trivial change. Do you want to take look?

flashdesignory commented:
> @rniwa @julienw I'm not sure if this qualifies as trivial change. Do you want to take look?

It does change scores and should get two approvals. 
Also @bgrins might be able to take a look since @julienw might not be able to get to this pr?

issackjohn commented:
@flashdesignory the Lit version of the TodoMVC sets the "completed" classname to the li like the rest of the frameworks
### todomvc-lit
```javascript
override render() {
        const itemClassList = {
            todo: true,
            completed: this.completed ?? false,
            editing: this.isEditing,
        };
        return html`
            <li class="${classMap(itemClassList)}">
```
 but the javascript-web-components TodoMVC uses an attribute "completed" in the custom component "todo-item" instead.
```js
toggleItem() {
        // The todo-list checks the "completed" attribute to filter based on route
        // (therefore the completed state needs to already be updated before the check)
        this.setAttribute("completed", this.toggleInput.checked);

        this.dispatchEvent(
            new CustomEvent("toggle-item", {
                detail: { id: this.id, completed: this.toggleInput.checked },
                bubbles: true,
            })
        );
    }
```

Is it possible to make either of these use the same approach as the other, this would help with the complex versions being able to share handcrafted CSS. Is this possible? Or what would you suggest?

flashdesignory commented:
> Is it possible to make either of these use the same approach as the other, this would help with the complex versions being able to share handcrafted CSS. Is this possible? Or what would you suggest?

The web component version is the only workload so far that is built according to my proposed css / a11y changes and therefore the structure is different from the other workloads. We are postponing making the same changes to the other todomvc apps and saving it for the next version (S4). 

If you want to make a change in the web component version, please add comments so that we can revert the changes at a later point.  Also, I would prefer not to touch the `/todomvc-css` folder, since that's set up for the new versions already.

julienw commented on resources/todomvc/big-dom-generator/gen-css.js:
```diff
@@ -152,6 +154,7 @@ const getInitialDepth = (element) => {
     return depth;
 };
 
+const cssProperties = [...CSS_PROPERTIES];
```
It's not clear to me why we're not using CSS_PROPERTIES directly.

That said I believe this file is removed in the other patch, so these changes won't be used here.

julienw commented on resources/todomvc/big-dom-generator/utils/buildComplex.js:
```diff
@@ -7,7 +7,7 @@ const COMPLEX_DOM_HTML_FILE = "index.html";
 const CSS_FILES_TO_ADD_LINKS_FOR = ["big-dom-generator.css", "generated.css"];
 
-function buildComplex(CALLER_DIRECTORY, SOURCE_DIRECTORY, TITLE, FILES_TO_MOVE) {
+function buildComplex(CALLER_DIRECTORY, SOURCE_DIRECTORY, TITLE, FILES_TO_MOVE, EXTRA_CSS_TO_LINK = [], SCRIPTS_TO_LINK = []) {
```
This will need to be updated with the new way of passing parameters in the other patch.

julienw reviewed (commented):
Just a few early comments. I think this will need to be updated after #291 anyway, right?

issackjohn commented:
> > Is it possible to make either of these use the same approach as the other, this would help with the complex versions being able to share handcrafted CSS. Is this possible? Or what would you suggest?
> 
> The web component version is the only workload so far that is built according to my proposed css / a11y changes and therefore the structure is different from the other workloads. We are postponing making the same changes to the other todomvc apps and saving it for the next version (S4).
> 
> If you want to make a change in the web component version, please add comments so that we can revert the changes at a later point. Also, I would prefer not to touch the `/todomvc-css` folder, since that's set up for the new versions already.

Okay that makes sense. Thanks!

lpardosixtosMs commented:
@flashdesignory @julienw. I updated the workload to work with handcrafted rules instead of a random generator. Can you take another look? I also updated the description and the workload design document. Here are the most recent perf numbers:

Before:

OS: macos 13.4.1 arm64. Device: MacBookPro17,1. CPU: Apple M1 8 cores.

**TodoMVC-Lit-Complex-DOM/total**
Firefox:  38.2 ± 3.1%
Chrome: 29.65 ± 0.65%
Safari: 62.76 ± 1.4%

**TodoMVC-Lit/total**
Firefox: 20.49 ± 1.4%
Chrome: 16.98 ± 0.61%
Safari: 21.16 ± 2.8%

**TodoMVC-WebComponents-Complex/total**
Firefox: 45.52 ± 1.4%
Chrome: 27.73 ± 0.48%
Safari 70.23 ± 1.1%

**TodoMVC-WebComponents/total**
Firefox: 28.13 ± 1.9%
Chrome: 23.79 ± 0.45% 
Safari: 23.39 ± 3.1%

**Sanity check. Numbers before this PR:**

**TodoMVC-Lit/total**
Firefox: 21.64 ± 3.0%
Chrome: 17.280 ± 0.29%
Safari: 21.30 ± 3.0%

**TodoMVC-WebComponents/total**
Firefox: 29.87 ± 0.46%
Chrome: 23.478 ± 0.33% 
Safari: 23.79 ± 1.6%

lpardosixtosMs commented:
> @lpardosixtosMs @issackjohn - is the requested change from @issackjohn completed and resolved?

Yes, but he's been out of office, that's why it is still there.

flashdesignory commented on resources/todomvc/big-dom-generator/utils/web-components-css/app.css:
```diff
@@ -0,0 +1,13 @@
+.todo-area {
+    font: 14px "Helvetica Neue", Helvetica, Arial, sans-serif;
+    line-height: 1.4em;
+    /* background: #f5f5f5; */
```
can we remove this?

lpardosixtosMs replied:
Done.

flashdesignory commented on resources/todomvc/big-dom-generator/utils/web-components-css/app.css:
```diff
@@ -0,0 +1,13 @@
+    min-width: 230px;
+    max-width: 550px;
+    margin: 0 10%;
+    /* margin: 0; */
```
can we remove this?

lpardosixtosMs replied:
Done.

flashdesignory commented on resources/todomvc/vanilla-examples/javascript-web-components-complex/a.out:
is this used for anything?

flashdesignory reviewed (approved):
some minor cleanup nits

julienw commented on resources/todomvc/big-dom-generator/utils/web-components-css/vanilla/todo-item-extra-css.js:
```diff
@@ -0,0 +1,40 @@
+const additionalStyleSheet = new CSSStyleSheet();
```
most comments I made above also apply here

julienw reviewed (commented):
I left a few comments, tell me what you think!

flashdesignory commented on resources/benchmark-runner.mjs:
makes sense - same

flashdesignory commented on resources/todomvc/big-dom-generator/utils/web-components-css/lit/todo-item-extra-css.js:
```diff
@@ -0,0 +1,40 @@
+    .todo > div label {
+        color: var(--complex-todo-red-label, --complex-label-color-default);
+    }
+    .todo.completed label {
```
The web component version uses a new css rule set that implements a11y updates that I discovered and presented. 
The rest of the todomvc workloads will get updated in S4.

lpardosixtosMs commented on resources/todomvc/architecture-examples/lit/src/lib/todo-item.ts:
```diff
@@ -7,6 +7,7 @@ import { classMap } from "lit/directives/class-map.js";
 import { todoStyles } from "./todo.css.js";
 import { DeleteTodoEvent, EditTodoEvent } from "./events.js";
 
+const EXTRA_CSS_TO_ADOPT = window.extraCssToAdopt;
```
Updated the name.

lpardosixtosMs commented on resources/todomvc/architecture-examples/lit/src/lib/todo-item.ts:
```diff
@@ -131,15 +132,25 @@ export class TodoItem extends LitElement {
 
+    override connectedCallback() {
+        super.connectedCallback();
+        if (EXTRA_CSS_TO_ADOPT)
```
Done.

lpardosixtosMs commented on resources/todomvc/architecture-examples/lit/src/lib/todo-item.ts:
```diff
@@ -181,10 +192,13 @@ export class TodoItem extends LitElement {
         input.value = this.text ?? "";
     }
 }
-
```
Done.

lpardosixtosMs commented on resources/todomvc/big-dom-generator/utils/web-components-css/lit/todo-item-extra-css.js:
```diff
@@ -0,0 +1,40 @@
+    priorityRules.push(`
+        :host([data-priority="${i}"]) > li {
+            --priority-background-color: var(--complex-todo-red-pri-${i});
+            --priority-background-color-completed: var(--complex-todo-green-pri-${i});
```
It is exactly for that, I added a comment.

lpardosixtosMs commented on resources/todomvc/big-dom-generator/utils/web-components-css/lit/todo-item-extra-css.js:
```diff
@@ -0,0 +1,40 @@
+priorityRules = priorityRules.join("\n");
+
+additionalStyleSheet.replaceSync(`
+    ${priorityRules}
```
Done.

lpardosixtosMs commented:
Addressed the feedback, Thanks!

@flashdesignory  @julienw I realized that adding the extra style sheet in the `connectedCallback` didn't make sense and could cause extra overhead when connecting the todo items. I refactored it so it is added either in the constructor (web components) or as part of the `styles` variable override (LIT).

@rniwa this change doesn't enable any test by default, do you want to take a look before merging?

rniwa reviewed (approved):
Looks sane to me.

flashdesignory commented:
> Addressed the feedback, Thanks!
> 
> @flashdesignory @julienw I realized that adding the extra style sheet in the `connectedCallback` didn't make sense and could cause extra overhead when connecting the todo items. I refactored it so it is added either in the constructor (web components) or as part of the `styles` variable override (LIT).
> 

@lpardosixtosMs - these changes are only related to complex dom styles, correct?

lpardosixtosMs commented:
> > Addressed the feedback, Thanks!
> > @flashdesignory @julienw I realized that adding the extra style sheet in the `connectedCallback` didn't make sense and could cause extra overhead when connecting the todo items. I refactored it so it is added either in the constructor (web components) or as part of the `styles` variable override (LIT).
> 
> @lpardosixtosMs - these changes are only related to complex dom styles, correct?

@flashdesignory correct. In the standalone case the extra stylesheet is not added (Web components) or an empty style entry is added (LIT).

julienw commented on resources/todomvc/big-dom-generator/utils/web-components-css/todo-list-extra-css.js:
```diff
@@ -0,0 +1,18 @@
+    --complex-todo-red-pri-0: rgb(253, 204, 204);
+    --complex-todo-red-pri-1: rgb(248, 221, 221);
+    --complex-todo-red-pri-2: rgb(251, 234, 234);
+    --complex-todo-red-pri-3: rgb(250, 240, 240);
```
can you please add a comment that pri-4 is missing on purpose?

julienw commented on resources/todomvc/big-dom-generator/utils/web-components-css/todo-list-extra-css.js:
```diff
@@ -0,0 +1,18 @@
+    --complex-todo-green-pri-0: rgb(204, 253, 204);
+    --complex-todo-green-pri-1: rgb(221, 248, 221);
+    --complex-todo-green-pri-2: rgb(234, 251, 234);
+    --complex-todo-green-pri-3: rgb(241, 250, 240);
```
same here

lpardosixtosMs replied:
Done.

julienw reviewed (approved):
Looks good to me, thanks
